std/gkrapi/compile.go

```diff
@@ -5,236 +5,282 @@ import (
+		opt(&res)
 	}
-	api.assignments = append(api.assignments, assignment)
-	return gkr.Variable(api.toStore.NewInputVariable()), nil
-}
 
-// appendNonNil filters out nil values from src and appends the non-nil values to dst.
-// i.e. dst = [0,1], src = [nil, 2, nil, 3] => dst = [0,1,2,3].
-func appendNonNil(dst *[]frontend.Variable, src []frontend.Variable) {
-	for i := range src {
-		if src[i] != nil {
-			*dst = append(*dst, src[i])
+	notOut := make([]bool, len(res.toStore.Circuit))
+	for i := range res.toStore.Circuit {
+		if res.toStore.Circuit[i].IsInput() {
+			res.ins = append(res.ins, gkr.Variable(i))
+		}
+		for _, inWI := range res.toStore.Circuit[i].Inputs {
+			notOut[inWI] = true
 		}
 	}
-}
-
-// Solve finalizes the GKR circuit and returns the output variables in the order created
-func (api *API) Solve(parentApi frontend.API) (Solution, error) {
 
-	var p gkrinfo.Permutations
-	var err error
-	if p, err = api.toStore.Compile(api.assignments.NbInstances()); err != nil {
-		return Solution{}, err
+	if len(res.ins) == len(res.toStore.Circuit) {
+		return nil, errors.New("circuit has no non-input wires")
 	}
-	api.assignments.Permute(p)
 
-	nbInstances := api.toStore.NbInstances
-	circuit := api.toStore.Circuit
+	for i := range res.toStore.Circuit {
+		if !notOut[i] {
+			res.outs = append(res.outs, gkr.Variable(i))
+		}
+	}
 
-	solveHintNIn := 0
-	solveHintNOut := 0
+	res.api.Compiler().Defer(res.finalize)
 
-	for i := range circuit {
-		v := &circuit[i]
-		in, out := v.IsInput(), v.IsOutput()
-		if in && out {
-			return Solution{}, fmt.Errorf("unused input (variable #%d)", i)
-		}
+	return &res, nil
+}
 
-		if in {
-			solveHintNIn += nbInstances - len(api.toStore.Dependencies[i])
-		} else if out {
-			solveHintNOut += nbInstances
+// AddInstance adds a new instance to the GKR circuit, returning the values of output variables for the instance.
+func (c *Circuit) AddInstance(input map[gkr.Variable]frontend.Variable) (map[gkr.Variable]frontend.Variable, error) {
+	if len(input) != len(c.ins) {
+		for k := range input {
+			if k >= gkr.Variable(len(c.toStore.Circuit)) {
+				return nil, fmt.Errorf("variable %d is out of bounds (max %d)", k, len(c.toStore.Circuit)-1)
+			}
+			if !c.toStore.Circuit[k].IsInput() {
+				return nil, fmt.Errorf("value provided for non-input variable %d", k)
+			}
 		}
 	}
-
-	// arrange inputs wire first, then in the order solved
-	ins := make([]frontend.Variable, 0, solveHintNIn)
-	for i := range circuit {
-		if circuit[i].IsInput() {
-			appendNonNil(&ins, api.assignments[i])
+	hintIn := make([]frontend.Variable, 2+len(c.ins)) // first and second input denote the circuit and instance, respectively.
+	hintIn[0] = c.index
+	hintIn[1] = c.toStore.NbInstances
+	for hintInI, wI := range c.ins {
+		if inV, ok := input[wI]; !ok {
+			return nil, fmt.Errorf("missing entry for input variable %d", wI)
+		} else {
+			hintIn[hintInI+2] = inV
+			c.assignments[wI] = append(c.assignments[wI], inV)
```

Comment:
Could we potentially have a perf regression here? If we add instances one-by-one then we append there the slices also one by one. If we have 50k instances, then it just adds up time- and allocation wise.

Thats why it would be nice to keep the previous approach as well -- then we know beforehand how big are the assignemnts and allocate only once. And we could also do solving only once?